webtau-http/src/main/java/com/twosigma/webtau/http/listener/HttpListener.java

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2019 TWO SIGMA OPEN SOURCE, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.twosigma.webtau.http.listener;
+
+import com.twosigma.webtau.http.HttpHeader;
+import com.twosigma.webtau.http.HttpResponse;
+import com.twosigma.webtau.http.request.HttpRequestBody;
+
+public interface HttpListener {
+    default void beforeFirstHttpCall() {
+    }
+
+    default void beforeHttpCall(String requestMethod,
+                                String passedUrl,
+                                String fullUrl,
+                                HttpHeader requestHeader,
```

Comment:
`HttpHeader` is still mutable, should we always pass in a copy to be safe?

https://github.com/twosigma/webtau/blob/master/webtau-http/src/main/java/com/twosigma/webtau/http/HttpHeader.java#L91

Reply:
Good question. Since we deprecated the call, I think it is fare to leave the responsibility on the person calling the deprecated method?

Reply:
That is true.  Perhaps we should also carry out the promise from the deprecation comment and actually start throwing from that function.

Reply:
Good idea. I will add throw. Will we have to increase the major version?

Reply:
Technically but I'm not sure we really need to at this stage.  @tsiq-clemens do you use `HttpHeader.add` anywhere in your tests?

Reply:
we merge headers at times, but I didn't see an `add` call